message/validation/validation.go

```diff
@@ -118,77 +121,105 @@ func (mv *messageValidator) handlePubsubMessage(pMsg *pubsub.Message, receivedAt
 	}
 
-	committee, validatorIndices, err := mv.getCommitteeAndValidatorIndices(signedSSVMessage.SSVMessage.GetID())
+	committeeInfo, err := mv.getCommitteeAndValidatorIndices(signedSSVMessage.SSVMessage.GetID())
 	if err != nil {
-		return nil, err
+		return decodedMessage, err
 	}
 
-	if err := mv.belongsToCommittee(signedSSVMessage.GetOperatorIDs(), committee); err != nil {
-		return nil, err
+	if err := mv.committeeChecks(signedSSVMessage, committeeInfo, topic); err != nil {
+		return decodedMessage, err
 	}
 
 	validationMu := mv.obtainValidationLock(signedSSVMessage.SSVMessage.GetID())
 
 	validationMu.Lock()
 	defer validationMu.Unlock()
 
-	decodedMessage := &queue.DecodedSSVMessage{
-		SignedSSVMessage: signedSSVMessage,
-		SSVMessage:       signedSSVMessage.SSVMessage,
-	}
-
 	switch signedSSVMessage.SSVMessage.MsgType {
 	case spectypes.SSVConsensusMsgType:
-		consensusMessage, err := mv.validateConsensusMessage(signedSSVMessage, committee, validatorIndices, receivedAt)
+		consensusMessage, err := mv.validateConsensusMessage(signedSSVMessage, committeeInfo, receivedAt)
+		decodedMessage.Body = consensusMessage
 		if err != nil {
-			return nil, err
+			return decodedMessage, err
 		}
 
-		decodedMessage.Body = consensusMessage
 	case spectypes.SSVPartialSignatureMsgType:
-		partialSignatureMessages, err := mv.validatePartialSignatureMessage(signedSSVMessage, committee, validatorIndices, receivedAt)
+		partialSignatureMessages, err := mv.validatePartialSignatureMessage(signedSSVMessage, committeeInfo, receivedAt)
+		decodedMessage.Body = partialSignatureMessages
 		if err != nil {
-			return nil, err
+			return decodedMessage, err
 		}
 
-		decodedMessage.Body = partialSignatureMessages
 	default:
-		panic("message type assertion should have been done")
+		panic("unreachable: message type assertion should have been done")
 	}
 
 	return decodedMessage, nil
 }
 
+func (mv *messageValidator) committeeChecks(signedSSVMessage *spectypes.SignedSSVMessage, committeeInfo CommitteeInfo, topic string) error {
+	if err := mv.belongsToCommittee(signedSSVMessage.GetOperatorIDs(), committeeInfo.operatorIDs); err != nil {
+		return err
+	}
+
+	// Rule: Check if message was sent in the correct topic
+	messageTopics := commons.CommitteeTopicID(committeeInfo.committeeID)
+	topicBaseName := commons.GetTopicBaseName(topic)
+	if !slices.Contains(messageTopics, topicBaseName) {
+		e := ErrIncorrectTopic
+		e.got = fmt.Sprintf("topic %v / base name %v", topic, topicBaseName)
+		e.want = messageTopics
+		return e
+	}
+
+	return nil
+}
+
 func (mv *messageValidator) obtainValidationLock(messageID spectypes.MessageID) *sync.Mutex {
 	// Lock this SSV message ID to prevent concurrent access to the same state.
 	mv.validationMutex.Lock()
 	mutex, ok := mv.validationLocks[messageID]
 	if !ok {
 		mutex = &sync.Mutex{}
 		mv.validationLocks[messageID] = mutex
+		// TODO: Clean the map when mutex won't be needed anymore. Now it's a mutex leak...
```

Comment:
we could use the ttlcache package i think we already import, however if we make sure to not instantiate this before verifying the committee/pk & role are correct, then we should not have more than `NumValidators * NumRoles` records in there (so at least we have an upper bound that should prevent mem leaks, and we can estimate the max memory usage to determine if TTL is even worth the effort)

Reply:
@moshe-blox Let's change this in another PR because this PR just adds a TODO?